server/src/main/java/org/elasticsearch/reservedstate/service/AbstractFileWatchingService.java

```diff
@@ -0,0 +1,392 @@
+            }
+        } catch (ClosedWatchServiceException | InterruptedException expected) {
+            logger.info("shutting down watcher thread");
+        } catch (Exception e) {
+            logger.error("shutting down watcher thread with exception", e);
+        }
+    }
+
+    final synchronized void stopWatcher() {
+        if (watching()) {
+            logger.debug("stopping watcher ...");
+            // make sure watch service is closed whatever
+            // this will also close any outstanding keys
+            try (var ws = watchService) {
+                watcherThread.interrupt();
+                watcherThread.join();
+
+                // make sure any keys are closed - if watchService.close() throws, it may not close the keys first
+                if (configDirWatchKey != null) {
+                    configDirWatchKey.cancel();
+                }
+                if (settingsDirWatchKey != null) {
+                    settingsDirWatchKey.cancel();
+                }
+            } catch (IOException e) {
+                logger.warn("encountered exception while closing watch service", e);
+            } catch (InterruptedException interruptedException) {
+                logger.info("interrupted while closing the watch service", interruptedException);
+            } finally {
+                watcherThread = null;
+                settingsDirWatchKey = null;
+                configDirWatchKey = null;
+                watchService = null;
+                logger.info("watcher service stopped");
+            }
+        } else {
+            logger.trace("file watch service already stopped");
+        }
+    }
+
+    // package private for testing
+    final WatchKey enableDirectoryWatcher(WatchKey previousKey, Path settingsDir) throws IOException, InterruptedException {
+        if (previousKey != null) {
+            previousKey.cancel();
+        }
+        int retryCount = 0;
+
+        do {
+            try {
+                return settingsDir.register(
+                    watchService,
+                    StandardWatchEventKinds.ENTRY_MODIFY,
+                    StandardWatchEventKinds.ENTRY_CREATE,
+                    StandardWatchEventKinds.ENTRY_DELETE
+                );
+            } catch (IOException e) {
+                if (retryCount == REGISTER_RETRY_COUNT - 1) {
+                    throw e;
+                }
+                Thread.sleep(retryDelayMillis(retryCount));
+                retryCount++;
+            }
+        } while (true);
+    }
+
+    // package private for testing
+    void processSettingsAndNotifyListeners() {
+        try {
+            processFileChanges();
+            for (var listener : eventListeners) {
+                listener.watchedFileChanged();
+            }
+        } catch (IOException | ExecutionException e) {
+            logger.error(() -> "Error processing watched file: " + watchedFile(), e.getCause());
+        } catch (InterruptedException e) {
```

Comment:
We should declare processSettingsAndNotifyListeners throws InterruptedException so this can be propagated as is. There is already a catch for it in the main watcherThread() method.

Reply:
Ah, I see. Thanks!